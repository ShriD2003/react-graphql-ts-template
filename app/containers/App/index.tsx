/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */
import React, { useState, useCallback, useEffect } from 'react';
import GlobalStyle from '@app/global-styles';
import { routeConfig } from '@app/routeConfig';
import { Layout, Drawer, DrawerProps } from 'antd';
import map from 'lodash-es/map';
import { withRouter } from 'react-router';
import { Link, Route, Switch } from 'react-router-dom';
import { compose } from 'redux';
import styled, { ThemeProvider } from 'styled-components';
import For from '@components/For';
import Header from '@components/Header';
import { colors } from '@themes/index';
import icon from '@images/ion_rocket-sharp.svg';
import { CloseOutlined } from '@ant-design/icons';

const theme = {
  fg: colors.primary,
  bg: colors.secondaryText
};

const SidebarDrawer = styled(Drawer)`
  && {
    .ant-drawer-body {
      padding: 7rem 0 0 0;
      background-color: ${colors.primary};
      width: 6%;
      min-width: 4.5rem;
      max-width: 7rem;
      text-align: center;
    }
    .ant-drawer-close {
      top: 1rem;
    }
  }
`;

const SideBar = styled.div`
  && {
    width: 6%;
    min-width: 4.5rem;
    max-width: 7rem;
    min-height: calc(100vh - 7rem);
    height: auto;
    background-color: ${colors.primary};
    display: inline;
    text-align: center;
  }
`;

const RocketLogo = styled.img`
  && {
    margin-top: 1rem;
    object-fit: contain;
  }
`;

const CustomLayout = styled(Layout)`
  && {
    flex-direction: row;
  }
`;

export function App() {
  const [visible, setVisible] = useState(false);
  const [mobile, setMobile] = useState(window.innerWidth < 400);

  function detectMobile() {
    if (window.innerWidth < 400) {
      setMobile(true);
    } else {
      setMobile(false);
    }
  }

  useEffect(() => {
    window.addEventListener('resize', detectMobile);
    return () => {
      window.removeEventListener('resize', detectMobile);
    };
  }, []);

  const toggleSidebar = useCallback(() => setVisible((v) => !v), []);

  const sidebarProps: DrawerProps = mobile
    ? {
        closeIcon: <CloseOutlined style={{ color: colors.secondary, fontSize: '1.9rem' }} />,
        placement: 'left',
        visible,
        closable: true,
        onClose: toggleSidebar,
        width: 'max-content'
      }
    : {};

  const SidebarComponent = (props: DrawerProps) =>
    mobile ? <SidebarDrawer {...props} /> : <SideBar data-testid="sidebar" {...(props as any)} />;

  return (
    <ThemeProvider theme={theme}>
      <Header mobile={mobile} toggleSidebar={toggleSidebar} />
      <CustomLayout>
        <SidebarComponent {...sidebarProps}>
          <Link data-testid="rocket-home-link" aria-label="home link" to="/">
            <RocketLogo src={icon} alt="rocket-icon" />
          </Link>
        </SidebarComponent>
        <Layout.Content>
          <For
            ParentComponent={(props) => <Switch {...props} />}
            of={map(Object.keys(routeConfig))}
            renderItem={(routeKey, index) => {
              const Component = routeConfig[routeKey].component;
              return (
                <Route
                  exact={routeConfig[routeKey].exact}
                  key={index}
                  path={routeConfig[routeKey].route}
                  render={(props) => {
                    const updatedProps = {
                      ...props,
                      ...routeConfig[routeKey].props
                    };
                    return <Component {...updatedProps} />;
                  }}
                />
              );
            }}
          />
          <GlobalStyle />
        </Layout.Content>
      </CustomLayout>
    </ThemeProvider>
  );
}

export default compose(withRouter)(App);
