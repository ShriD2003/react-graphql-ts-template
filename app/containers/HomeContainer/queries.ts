import { gql } from 'apollo-boost';

export const GET_LAUNCHES = gql`
  query launches($missionName: String) {
    launches(find: { mission_name: $missionName }) {
      launch_date_local
      mission_name
      links {
        wikipedia
        flickr_images
      }
    }
  }
`;

export interface LaunchesQueryVariables {
  missionName: string;
}
