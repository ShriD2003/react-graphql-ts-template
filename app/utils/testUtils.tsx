import React from 'react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import history from './history';

export const renderWithIntl = (children: React.ReactNode) => {
  i18n.load(DEFAULT_LOCALE, translationMessages[DEFAULT_LOCALE]);
  i18n.activate(DEFAULT_LOCALE);

  return render(<I18nProvider i18n={i18n}> {children} </I18nProvider>);
};

export const getComponentStyles = (Component: React.FC<any>, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props)!.type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  return window.getComputedStyle(componentRoots[0]);
};

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});
