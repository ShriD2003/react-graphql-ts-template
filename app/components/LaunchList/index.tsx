import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { get, isEmpty } from 'lodash-es';
import { Card, Skeleton } from 'antd';
import { If, T, For } from '@components';
import { colors } from '@app/themes';
import { Launch } from '@app/containers/HomeContainer';

const CustomCard = styled(Card)`
  && {
    color: ${colors.primary};
    background-color: ${colors.secondaryText};
  }
`;
const WrapperCard = styled(Card)`
  && {
    color: ${colors.primary};
    border: none;
    background-color: ${colors.secondaryText};
  }
`;

const CustomErrorCard = styled(Card)`
  && {
    color: ${colors.secondary};
    margin: 2rem;
    background-color: ${colors.secondaryText};
  }
`;

const Container = styled.div`
  && {
    display: flex;
    flex-direction: column;
    width: 100%;
    margin: 0 auto;
    background-color: ${colors.secondaryText};
  }
`;

interface LaunchListProps {
  launchData: { launches?: Launch[] };
  loading: boolean;
}

export function LaunchList({ launchData, loading }: LaunchListProps) {
  const launches = get(launchData, 'launches', []);
  return (
    <If
      condition={!isEmpty(launches) || loading}
      otherwise={
        <CustomErrorCard>
          <T data-testid="default-message" id="fallback" />
        </CustomErrorCard>
      }
    >
      <WrapperCard data-testid="list">
        <Skeleton loading={loading} active>
          <For
            of={launches}
            ParentComponent={Container}
            renderItem={(launch: Launch, idx) => (
              <CustomCard key={idx}>
                <div>{launch.mission_name}</div>
                <div> {launch.launch_date_local}</div>
              </CustomCard>
            )}
          />
        </Skeleton>
      </WrapperCard>
    </If>
  );
}

LaunchList.propTypes = {
  launchData: PropTypes.shape({
    launches: PropTypes.arrayOf(
      PropTypes.shape({
        mission_name: PropTypes.string,
        launch_date_local: PropTypes.string,
        links: PropTypes.shape({
          wikipedia: PropTypes.string,
          flick_images: PropTypes.array
        })
      })
    )
  }),
  loading: PropTypes.bool
};

export default memo(LaunchList);
