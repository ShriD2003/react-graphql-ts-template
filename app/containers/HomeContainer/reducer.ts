import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash-es/get';
import { Launch } from '@app/containers/HomeContainer';

export type HomeContainerState = {
  launchData: {
    launches?: Launch[];
  };
  launchListError: any;
  loading: boolean;
};

export const { Types: homeContainerTypes, Creators: homeContainerCreators } = createActions({
  requestGetLaunchList: { missionName: null, order: 'asc', page: 1 },
  successGetLaunchList: ['launchData'],
  failureGetLaunchList: ['launchListError']
});

export const initialState: HomeContainerState = {
  loading: false,
  launchData: {},
  launchListError: null
};

export interface HomeContainerAction extends Partial<Omit<HomeContainerState, 'loading'>> {
  type?: string;
  order?: string;
  page?: number;
  missionName?: string;
}

export const homeContainerReducer = (state = initialState, action: HomeContainerAction) =>
  produce(state, (draft) => {
    switch (action.type) {
      case homeContainerTypes.REQUEST_GET_LAUNCH_LIST:
        draft.loading = true;
        break;

      case homeContainerTypes.SUCCESS_GET_LAUNCH_LIST:
        draft.launchData = action.launchData!;
        draft.launchListError = null;
        draft.loading = false;
        break;
      case homeContainerTypes.FAILURE_GET_LAUNCH_LIST:
        draft.loading = false;
        draft.launchListError = draft.launchListError = get(action.launchListError, 'message', 'something_went_wrong');
        break;
    }
  });

export default homeContainerReducer;
