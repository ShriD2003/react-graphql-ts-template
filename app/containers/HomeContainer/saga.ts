import { put, call, takeLatest } from 'redux-saga/effects';
import { getLaunches } from '@services/launchApi';
import { homeContainerTypes, homeContainerCreators } from './reducer';
import { DefaultActionTypes } from 'reduxsauce';
import { launch } from '.';

const { REQUEST_GET_LAUNCH_LIST }: DefaultActionTypes = homeContainerTypes;
const { successGetLaunchList, failureGetLaunchList } = homeContainerCreators;
export function* getLaunchList(): Generator<any, any, { data: { launches: launch; errors: Object }; ok: boolean }> {
  const response = yield call(getLaunches);
  const { data, ok } = response;
  if (ok) {
    if (!data.errors) {
      yield put(successGetLaunchList(data));
    } else {
      yield put(failureGetLaunchList(data));
    }
  } else {
    yield put(failureGetLaunchList(data));
  }
}
// Individual exports for testing
export default function* homeContainerSaga() {
  yield takeLatest(REQUEST_GET_LAUNCH_LIST, getLaunchList);
}
