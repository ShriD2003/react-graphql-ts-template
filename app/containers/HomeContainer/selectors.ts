import { createSelector } from 'reselect';
import get from 'lodash/get';
import { initialState } from './reducer';

/**
 * Direct selector to the homeContainer state domain
 */

export const selectHomeContainerDomain = (state: any) => state.homeContainer || initialState;

/**
 * Other specific selectors
 */

/**
 * Default selector used by HomeContainer
 */

export const selectLaunchData: any = () =>
  createSelector(selectHomeContainerDomain, (substate) => get(substate, 'launchData'));

export const selectLaunchListError: any = () =>
  createSelector(selectHomeContainerDomain, (substate) => get(substate, 'launchListError'));

export const selectLaunchQuery: any = () =>
  createSelector(selectHomeContainerDomain, (substate) => get(substate, 'launchQuery'));
