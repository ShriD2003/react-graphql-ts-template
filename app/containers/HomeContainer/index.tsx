import React, { useEffect, memo } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { AnyAction, compose } from 'redux';
import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';
import styled from 'styled-components';
import { injectIntl, IntlShape } from 'react-intl';
import { injectSaga } from 'redux-injectors';
import { Card, Skeleton } from 'antd';
import If from '@components/If';
import { selectLaunchData, selectLaunchListError, selectLoading } from './selectors';
import { homeContainerCreators } from './reducer';
import homeContainerSaga from './saga';
import For from '@app/components/For';
import { ErrorHandler } from '@app/components/ErrorHandler';

const CustomCard = styled(Card)`
  && {
    margin: 20px 0;
    color: ${(props) => props.color};
  }
`;
const Container = styled.div`
  && {
    display: flex;
    flex-direction: column;
    width: 100%;
    margin: 0 auto;
  }
`;

export interface launch {
  mission_name: string;
  launch_date_local: string;
  links: {
    wikipedia: string;
    flick_images: Array<string>;
  };
}

interface HomeContainerProps {
  dispatchLaunchList: Function;
  dispatchClearLaunchList: Function;
  launchData: {
    data: launch;
  };
  launchListError: string;
  intl: IntlShape;
}

export function HomeContainer({
  dispatchLaunchList,
  dispatchClearLaunchList,
  intl,
  loading,
  launchData,
  launchListError
}: HomeContainerProps | any) {
  useEffect(() => {
    dispatchClearLaunchList();
    dispatchLaunchList();
  }, []);

  const renderLaunchList = () => {
    const launches = get(launchData, 'launches', []);
    return (
      <If condition={!isEmpty(launches) || loading}>
        <CustomCard data-testid="list">
          <Skeleton loading={loading} active>
            <For
              of={launches}
              ParentComponent={Container}
              renderItem={(launch: launch, idx) => (
                <CustomCard key={idx}>
                  <div>{launch.mission_name}</div>
                  <div> {launch.launch_date_local}</div>
                </CustomCard>
              )}
            ></For>
          </Skeleton>
        </CustomCard>
      </If>
    );
  };

  return (
    <Container>
      <CustomCard title={intl.formatMessage({ id: 'spacex_search' })} />
      {renderLaunchList()}
      <ErrorHandler loading={loading} launchListError={launchListError} />
    </Container>
  );
}

HomeContainer.propTypes = {
  dispatchLaunchList: PropTypes.func,
  dispatchClearLaunchList: PropTypes.func,
  intl: PropTypes.object,
  launchData: PropTypes.shape({
    totalCount: PropTypes.number,
    incompleteResults: PropTypes.bool,
    items: PropTypes.array
  }),
  launchListError: PropTypes.string,
  history: PropTypes.object,
  maxwidth: PropTypes.number,
  padding: PropTypes.number
};

HomeContainer.defaultProps = {
  maxwidth: 500,
  padding: 20,
  launchData: {},
  launchListError: null
};

const mapStateToProps = createStructuredSelector({
  launchData: selectLaunchData(),
  launchListError: selectLaunchListError(),
  loading: selectLoading()
});

export function mapDispatchToProps(dispatch: (arg0: AnyAction) => any) {
  const { requestGetLaunchList, clearLaunchList } = homeContainerCreators;
  return {
    dispatchLaunchList: () => dispatch(requestGetLaunchList()),
    dispatchClearLaunchList: () => dispatch(clearLaunchList())
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(
  injectIntl,
  withConnect,
  memo,
  injectSaga({ key: 'homeContainer', saga: homeContainerSaga })
)(HomeContainer);

export const HomeContainerTest = compose(injectIntl)(HomeContainer);
