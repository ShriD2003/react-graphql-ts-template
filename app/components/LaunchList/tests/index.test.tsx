/**
 *
 * Tests for LaunchList
 *
 */

import React from 'react';
import { renderWithIntl } from '@utils/testUtils';
import LaunchList from '../index';
import { LaunchData } from '@app/containers/HomeContainer';

describe('<LaunchList />', () => {
  const loading = false;
  const launchData: LaunchData = {
    launches: [
      {
        launchDateLocal: '2014-01-06T14:06:00-04:00',
        launchDateUnix: 123123123,
        missionName: 'Thaicom 6',
        links: {
          wikipedia: 'https://en.wikipedia.org/wiki/Thaicom_6',
          flickrImages: ['https://farm9.staticflickr.com/8617/16789019815_f99a165dc5_o.jpg']
        }
      }
    ]
  };
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<LaunchList loading={loading} launchData={launchData} />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should show the fallbackMessage if luanchData is empty', () => {
    const message = 'No results found for the search term.';
    const { getByText } = renderWithIntl(<LaunchList loading={loading} launchData={{}} />);
    expect(getByText(message)).toBeInTheDocument();
  });
  it('should render the list for the launches when data is available', () => {
    const { getByText } = renderWithIntl(<LaunchList loading={loading} launchData={launchData} />);
    expect(getByText('Thaicom 6')).toBeInTheDocument();
  });
});
