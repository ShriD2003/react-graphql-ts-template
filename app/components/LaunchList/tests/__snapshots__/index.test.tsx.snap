// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchList /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="LaunchList__Container-sc-1l6drbj-1 koCUuC"
      data-testid="for"
      orientation="row"
    >
      <div
        class="ant-card ant-card-bordered LaunchItem__LaunchCard-tig4gy-0 eDryXK"
      >
        <div
          class="ant-card-body"
        >
          <p
            class="T__StyledText-gcrjip-0 lijFEY"
            data-testid="mission-name"
          >
            Thaicom 6
          </p>
          <p
            class="T__StyledText-gcrjip-0 cSxVIW"
            data-testid="t"
          >
            2014-01-06T14:06:00-04:00
          </p>
          <a
            class="ant-btn ant-btn-link LaunchItem__WikiLink-tig4gy-1 fGRbFh"
            href="https://en.wikipedia.org/wiki/Thaicom_6"
            rel="noreferrer"
            target="_blank"
          >
            <span>
              Wiki Link
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
